### PR TITLE
Enhancement/8560 new campaign signal

### PR DESCRIPTION
**Pull request by zutigrm**
## Summary

Addresses issue:

- #8560

## Relevant technical choices

* Refactored approach a bit, by introducing the `PAX_SETUP_ENUM` which determines if we should either render the PAX app, or hide it and render the `complete setup` button. It looks better to render button on the end, then having it stick around in disabled state. We should probably consider including some kind of Campaign success setup message in the near future, but this might be covered in the design once done. As on the final step, screen remains on the campaign overview, there is no other message type of screen in the PAX itself, so it looks a bit weird if we don't hide it
* `PAXEmbeddedApp` does not have tests implemented as it is only a wrapper.  Services tests are handled in separate file.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/modules/ads/components/setup/SetupMainPAX.js [hunk 1]
```diff
@@ -79,16 +93,20 @@ export default function SetupMainPAX( { finishSetup } ) {
...
+		if ( ! paxApp.current ) {
 			return;
 		}
 
 		/* eslint-disable sitekit/acronym-case */
 		// Disabling rule because function and property names
 		// are expected in current format by PAX API.
 		const { accountService, conversionTrackingIdService } =
-			paxApp.getServices();
+			paxApp.current.g;
```
Comment (zutigrm) [inline assets/js/modules/ads/components/setup/SetupMainPAX.js:109]: Apparently there is no `getServices()` anymore in passed `app` object, it became `g`

Comment (zutigrm) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js]: Solved by changing integrator version to the `latest`

File: assets/js/modules/ads/pax/services.test.js [hunk 2]
```diff
@@ -73,7 +73,7 @@ describe( 'PAX partner services', () => {
 							},
 						},
 					};
-					services = createPaxServices( registry, {
+					services = createPaxServices( registry, null, {
```
Comment (tofumatt) [inline assets/js/modules/ads/pax/services.test.js:76]: Having to supply `null` as an argument like this isn't a great API. Let's change the second argument to be an object instead, since the arguments are all optional. We shouldn't have optional positional arguments 🙂

File: assets/js/modules/ads/components/setup/SetupMainPAX.js [hunk 3]
```diff
@@ -79,16 +97,31 @@ export default function SetupMainPAX( { finishSetup } ) {
...
+		}
+	} );
+
+	const onLaunch = ( app ) => {
+		if ( ! paxAppRef.current ) {
+			paxAppRef.current = app;
+		}
+	};
+
+	const onCampaignCreated = useCallbackOne( async () => {
```
Comment (tofumatt) [inline assets/js/modules/ads/components/setup/SetupMainPAX.js:115]: Why do we need `useCallbackOne` here instead of `useCallback`?

File: assets/js/modules/ads/components/setup/SetupMainPAX.js [hunk 4]
```diff
@@ -79,16 +97,31 @@ export default function SetupMainPAX( { finishSetup } ) {
...
+			// set the step back to the PAX launch, as values are only temporarly
+			// saved in state after PAX campaign setup signal is received.
+			setShowPaxAppQueryParam( PAX_SETUP_STEP.LAUNCH );
+		}
+	} );
+
+	const onLaunch = ( app ) => {
+		if ( ! paxAppRef.current ) {
+			paxAppRef.current = app;
+		}
```
Comment (tofumatt) [inline assets/js/modules/ads/components/setup/SetupMainPAX.js:112]: Do we need to check for an existing value here? Presumably this will only run once, but it's also safe to reassign this I'd think 🤔 

```suggestion
		paxAppRef.current = app;
```

**Review by tofumatt — CHANGES_REQUESTED** [review]
Just a few small questions/comments here, but largely looks good 🙂

File: assets/js/modules/ads/components/setup/SetupMainPAX.js [hunk 5]
```diff
@@ -79,16 +94,31 @@ export default function SetupMainPAX( { finishSetup } ) {
 	const { setPaxConversionID, setExtCustomerID, submitChanges } =
 		useDispatch( MODULES_ADS );
 
-	const onCompleteSetup = useCallback( async () => {
-		if ( ! paxApp ) {
+	useMount( () => {
+		if ( PAX_SETUP_ENUM.PAX_SETUP_FINISHED === +showPaxApp ) {
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js]: I will use `parseInt` for readability sake then

File: assets/js/modules/ads/pax/services.test.js (hunk 2 above)
Comment (zutigrm) [reply assets/js/modules/ads/pax/services.test.js:76]: Yeah that's a good point

File: assets/js/modules/ads/components/setup/SetupMainPAX.js (hunk 3 above)
Comment (zutigrm) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js:115]: Had a chat with Evan 2 days ago, he mentioned we might need ensure this callback does not cause re-render to the PAX app, so I used this approach to avoid instances where React might clear regular `useCallback` causing the re-render, as `useCallbackOne` will persist.

File: assets/js/modules/ads/components/setup/SetupMainPAX.js (hunk 4 above)
Comment (zutigrm) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js:112]: Due to some weirdness in PAX app, like `getServices()` not being callable, then working again, had to put few safeguards to ensure some unexpected changes won't happen

File: assets/js/modules/ads/components/setup/SetupMainPAX.js (hunk 3 above)
Comment (tofumatt) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js:115]: Ah, yes, good point. 👍🏻

File: assets/js/modules/ads/components/setup/SetupMainPAX.js (hunk 4 above)
Comment (tofumatt) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js:112]: I don't understand though, `onLaunch` is only ever called once, right?

If `paxAppRef.current` is defined, it means that this code was already called. What bug is the check for `! paxAppRef.current` fixing? This code should not be called if `paxAppRef.current` is set, because it is only set here and this code only runs once.

Can you include a screencast of the bug you experience when this change isn't made? 🤔

If it does need to stay we should leave a comment explaining it, because it's not intuitive that this code runs multiple times.

**Review by tofumatt — CHANGES_REQUESTED** [review]
Looks good, thanks for the changes made, I think it cleans things up a lot 👍🏻 

Just one more question about the `onLaunch` function seemingly being called multiple times, which is unexpected. Can you let me know what happens if we remove the `if ( ! paxAppRef.current )` check?

File: assets/js/modules/ads/components/setup/SetupMainPAX.js (hunk 4 above)
Comment (zutigrm) [reply assets/js/modules/ads/components/setup/SetupMainPAX.js:112]: Technically shouldn't make a problem to remove it, tested it now without it, not causing any unexpected errors. It was set as a local state before, when Simon gave it a quick test, app instance was null in campaign created callback, although `getServices()` was inaccessible even when I changed the approach. Anyway it was throwing different weirdness for that 1 day only, maybe it was some cache, or some delay on their server, etc. 

TLDR I removed it now, should be good

**Comment by aaemnnosttv** [comment]
> * Refactored approach a bit, by introducing the `PAX_SETUP_ENUM` which determines if we should either render the PAX app, or hide it and render the `complete setup` button. It looks better to render button on the end, then having it stick around in disabled state.

@zutigrm please refrain from making changes to the interface as part of an issue like this as it's out of scope. I agree it's a bit odd the way it is, but it's consistent with other modules and generally conditionally hiding/showing elements introduces shifting which we try to avoid. We don't need to undo it right now since we're going to revisit the flow soon anyways but please keep in mind going forward.
